### ISSUE TITLE
Fix bootloader build script for AIX

### ISSUE THREAD

**Issue by aixtools** [pyinstaller/pyinstaller#4211]
To get PyInstaller able to successfully build (and install) the bootloader I had to make a few changes to bootloader/wscript

One is simply a 'typo', the other - I am not as sure about - so I just commented it away.

Here is the diff -u (open for reactions before I assemble a PR)

```
HEAD is now at 0f31b35... Release 3.4.

diff -ru ./wscript /home/root/python/PyInstaller-3.4/bootloader/wscript
--- ./wscript   2019-05-04 09:50:41.000000000 +0000
+++ /home/root/python/PyInstaller-3.4/bootloader/wscript        2019-05-01 14:18:49.000000000 +0000
@@ -461,11 +461,11 @@
          ctx.find_program([assoc_programm(ctx, 'strip')], var='STRIP')
          ctx.load('strip', tooldir='tools')
          # There is a strip flag for AIX environment
-         if ctx.env.DEST_OS == 'aix':
-             if ctx.env.PYI_ARCH == '32bit':
-                 ctx.env.append_value('STRIPFLAGS', '-X32')
-             elif ctx.env.PYI_ARCH == '64bit':
-                 ctx.env.append_value('STRIPFLAGS', '-X64')
+         # if ctx.env.DEST_OS == 'aix':
+         #     if ctx.env.PYI_ARCH == '32bit':
+         #         ctx.env.append_value('STRIPFLAGS', '-X32')
+         #     elif ctx.env.PYI_ARCH == '64bit':
+         #         ctx.env.append_value('STRIPFLAGS', '-X64')

      ### C Compiler optimizations.
      # TODO Set proper optimization flags for MSVC (Visual Studio).
@@ -771,7 +771,7 @@
          if ctx.env.DEST_OS == 'aix':
              # link statically with zlib
              libs.remove('Z')
-             staticlibs.append('Z')
+             staticlibs.append('z')

          if ctx.options.boehmgc:
              libs.append('GC')
```

**Comment by htgoebel** [comment]
Well, AIX is a contributed platform only, so I can't tell much except:
* The STRIPFLAGS have been added intentionally somewhen. Why are they not longer needed? How to test whether they are needed?
* Regarding "Z": this seems to be intriduced in 47ca3e5622719610abfa4471b432822be5584bab and should be fixed.
I'd appreciate a pull-request containing two commits: one for each change. Thanks.

**Comment by aixtools** [comment]
Sorry I have been so long. Coming up.

Traditionally, stripflags has been to make executables smaller.

I'll try and get these done today.

**Comment by aixtools** [comment]
I had made a PR for the strip flags - that I closed, but would like to repoen to to issues with building the bootloader. With the flags being set I get an error such as:

[11/13] Compiling src/pyi_pythonlib.c
[12/13] Linking build/debug/run_d
[13/13] Processing build/debug/run_d
/usr/vac/bin/xlc_r: 1501-216 (W) command option -32 is not recognized - passed to ld
ld: 0706-012 The -X flag is not recognized.
ld: 0706-012 The -3 flag is not recognized.
ld: 0706-012 The -2 flag is not recognized.

So, apparently they are also being sent to CC (which was already set using -q32 (or -q64).

The 'normal' way (for myself) to be handling these things is to set (and export) OBJECT_MODE=32 or OBJECT_MODE=64 in the environment.

However, if you can help me out with finding where the **STRIPFLAG** is being passed to **CC** (that passes it to `/usr/bin/ld`) then I'll make a much better PR for setting X32 or X64 for strip.

**Comment by aixtools** [comment]
Thanks for merging #4555, will you reopen #4556 - as there is still something not working as expected. The flag is being passed to CC - and never gets far enough to pass it to strip.

**Comment by aixtools** [comment]
re: the call to strip - this seems to be going wrong in waflib. I modified ./bootloader/.waf3-2.0.19-9054d7a7f551c4b14b458abcd389b706/waflib/Task.py" to:

  +249          def format_error(self):
  +250                  if Logs.verbose:
  +251                          msg=': %r\n%r'%(self,getattr(self,'last_cmd',''))
  +252                  else:
  +253                          msg=': %r\n%r'%(self,getattr(self,'last_cmd',''))
  +254                          # msg=' (run with -v to display more information)'

and now get:
...
[10/13] Compiling src/pyi_path.c
[11/13] Compiling src/pyi_pythonlib.c
[12/13] Linking build/debug/run_d
[13/13] Processing build/debug/run_d
/usr/vac/bin/xlc_r: 1501-216 (W) command option -32 is not recognized - passed to ld
ld: 0706-012 The -X flag is not recognized.
ld: 0706-012 The -3 flag is not recognized.
ld: 0706-012 The -2 flag is not recognized.

Waf: Leaving directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
Build failed
 -> task in 'run_d' failed with exit status 255:
        {task 809033392: strip run_d -> }
['/usr/vac/bin/xlc_r', '-X32', 'run_d']
ERROR: Failed compiling the bootloader. Please compile manually and rerun setup.py

Hints welcome!

**Comment by aixtools** [comment]
After digging a bit deeper I found something in 

I made a small change that is still confusing me. With:
```
  +642      if ctx.env.CC_NAME != 'msvc':
  +643          # This tool allows reducing the size of executables.
  +644          if ctx.env.DEST_OS == 'aix':
  +645              ctx.find_program('strip')
  +646          else:
  +647              ctx.find_program([assoc_programm(ctx, 'strip')], var='STRIP')
  +648          ctx.load('strip', tooldir='tools')
  +649          # There is a strip flag for AIX environment
  +650          if ctx.env.DEST_OS == 'aix':
  +651              if ctx.env.PYI_ARCH == '32bit':
  +652                  ctx.env.append_value('STRIPFLAGS', '-X32')
  +653              elif ctx.env.PYI_ARCH == '64bit':
  +654                  ctx.env.append_value('STRIPFLAGS', '-X64')
  +655
```

What I do not understand - why is it checking for strip twice? And why is it doing the 'debug' directory twice (and then fails because it has already stripped in 'debug')

I get:
```
aixtools@x064:[/data/prj/python/git/pyinstaller]python ./setup.py build
running build
running build_bootloader
No precompiled bootloader found. Trying to compile it for you ...
Setting top to                           : /data/prj/python/git/pyinstaller/bootloader
Setting out to                           : /data/prj/python/git/pyinstaller/bootloader/build
Python Version                           : 3.7.4 (default, Dec 24 2019, 19:05:28) [C]
Checking for 'xlc' (C compiler)          : /usr/vac/bin/xlc_r
Checking size of pointer                 : 4
Platform                                 : AIX-32bit detected based on compiler
Checking for compiler flags -q32         : yes
Checking for library dl                  : yes
Checking for library m                   : yes
Checking for library z                   : yes
Checking for function unsetenv           : yes
Checking for function mkdtemp            : yes
Checking for function strndup            : yes
Checking for function strnlen            : yes
Checking for program 'strip'             : /usr/bin/strip
Checking for program 'strip'             : /usr/bin/strip
'configure' finished successfully (1.585s)
'all' finished successfully (0.001s)
'distclean' finished successfully (0.134s)
Setting top to                           : /data/prj/python/git/pyinstaller/bootloader
Setting out to                           : /data/prj/python/git/pyinstaller/bootloader/build
Python Version                           : 3.7.4 (default, Dec 24 2019, 19:05:28) [C]
Checking for 'xlc' (C compiler)          : /usr/vac/bin/xlc_r
Checking size of pointer                 : 4
Platform                                 : AIX-32bit detected based on compiler
Checking for compiler flags -q32         : yes
Checking for library dl                  : yes
Checking for library m                   : yes
Checking for library z                   : yes
Checking for function unsetenv           : yes
Checking for function mkdtemp            : yes
Checking for function strndup            : yes
Checking for function strnlen            : yes
Checking for program 'strip'             : /usr/bin/strip
Checking for program 'strip'             : /usr/bin/strip
'configure' finished successfully (1.567s)
'make_all' finished successfully (0.015s)
Waf: Entering directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
[ 1/13] Compiling src/pyi_python.c
[ 2/13] Compiling src/pyi_global.c
[ 3/13] Compiling src/main.c
[ 4/13] Compiling src/pyi_archive.c
[ 5/13] Compiling src/pyi_utils.c
[ 6/13] Compiling src/pyi_launch.c
[ 7/13] Compiling src/pyi_python27_compat.c
[ 8/13] Compiling src/pyi_pythonlib.c
[ 9/13] Compiling src/pyi_win32_utils.c
"../../src/pyi_win32_utils.c", line 574.21: 1506-356 (W) Compilation unit is empty.

[10/13] Compiling src/pyi_path.c
[11/13] Compiling src/pyi_main.c
[12/13] Linking build/debug/run_d
[13/13] Processing build/debug/run_d
Waf: Leaving directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
'build_debug' finished successfully (1.005s)
Waf: Entering directory `/data/prj/python/git/pyinstaller/bootloader/build/release'
[ 1/13] Compiling src/pyi_python.c
[ 2/13] Compiling src/pyi_global.c
[ 3/13] Compiling src/main.c
[ 4/13] Compiling src/pyi_archive.c
[ 5/13] Compiling src/pyi_utils.c
[ 6/13] Compiling src/pyi_launch.c
[ 7/13] Compiling src/pyi_python27_compat.c
[ 8/13] Compiling src/pyi_pythonlib.c
[ 9/13] Compiling src/pyi_win32_utils.c
"../../src/pyi_win32_utils.c", line 574.21: 1506-356 (W) Compilation unit is empty.

[10/13] Compiling src/pyi_path.c
[11/13] Compiling src/pyi_main.c
[12/13] Linking build/release/run
[13/13] Processing build/release/run
Waf: Leaving directory `/data/prj/python/git/pyinstaller/bootloader/build/release'
'build_release' finished successfully (0.919s)
Waf: Entering directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
[13/14] Processing build/debug/run_d
+ install /data/prj/python/git/pyinstaller/PyInstaller/bootloader/AIX-32bit/run_d (from build/debug/run_d)
strip: run_d -- 0654-420 The file was already stripped as specified.

Waf: Leaving directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
Build failed
 -> task in 'run_d' failed with exit status 255:
        {task 810096624: strip run_d -> }
['/usr/bin/strip', '-X32', 'run_d']
ERROR: Failed compiling the bootloader. Please compile manually and rerun setup.py

```
FYI: before the change I get: Note - still two "Checking for program..." lines, but for different program names.
```

aixtools@x064:[/data/prj/python/git/pyinstaller]python ./setup.py build
running build
running build_bootloader
No precompiled bootloader found. Trying to compile it for you ...
Setting top to                           : /data/prj/python/git/pyinstaller/bootloader
Setting out to                           : /data/prj/python/git/pyinstaller/bootloader/build
Python Version                           : 3.7.4 (default, Dec 24 2019, 19:05:28) [C]
Checking for 'xlc' (C compiler)          : /usr/vac/bin/xlc_r
Checking size of pointer                 : 4
Platform                                 : AIX-32bit detected based on compiler
Checking for compiler flags -q32         : yes
Checking for library dl                  : yes
Checking for library m                   : yes
Checking for library z                   : yes
Checking for function unsetenv           : yes
Checking for function mkdtemp            : yes
Checking for function strndup            : yes
Checking for function strnlen            : yes
Checking for program '/usr/vac/bin/xlc_r' : /usr/vac/bin/xlc_r
Checking for program 'strip'              : /usr/vac/bin/xlc_r
'configure' finished successfully (1.669s)
'all' finished successfully (0.001s)
'distclean' finished successfully (0.134s)
Setting top to                           : /data/prj/python/git/pyinstaller/bootloader
Setting out to                           : /data/prj/python/git/pyinstaller/bootloader/build
Python Version                           : 3.7.4 (default, Dec 24 2019, 19:05:28) [C]
Checking for 'xlc' (C compiler)          : /usr/vac/bin/xlc_r
Checking size of pointer                 : 4
Platform                                 : AIX-32bit detected based on compiler
Checking for compiler flags -q32         : yes
Checking for library dl                  : yes
Checking for library m                   : yes
Checking for library z                   : yes
Checking for function unsetenv           : yes
Checking for function mkdtemp            : yes
Checking for function strndup            : yes
Checking for function strnlen            : yes
Checking for program '/usr/vac/bin/xlc_r' : /usr/vac/bin/xlc_r
Checking for program 'strip'              : /usr/vac/bin/xlc_r
'configure' finished successfully (1.889s)
'make_all' finished successfully (0.015s)
Waf: Entering directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
[ 1/13] Compiling src/pyi_python.c
[ 2/13] Compiling src/pyi_global.c
[ 3/13] Compiling src/main.c
[ 4/13] Compiling src/pyi_archive.c
[ 5/13] Compiling src/pyi_utils.c
[ 6/13] Compiling src/pyi_launch.c
[ 7/13] Compiling src/pyi_python27_compat.c
[ 8/13] Compiling src/pyi_pythonlib.c
[ 9/13] Compiling src/pyi_win32_utils.c
[10/13] Compiling src/pyi_path.c
"../../src/pyi_win32_utils.c", line 574.21: 1506-356 (W) Compilation unit is empty.

[11/13] Compiling src/pyi_main.c
[12/13] Linking build/debug/run_d
[13/13] Processing build/debug/run_d
/usr/vac/bin/xlc_r: 1501-216 (W) command option -32 is not recognized - passed to ld
ld: 0706-012 The -X flag is not recognized.
ld: 0706-012 The -3 flag is not recognized.
ld: 0706-012 The -2 flag is not recognized.

Waf: Leaving directory `/data/prj/python/git/pyinstaller/bootloader/build/debug'
Build failed
 -> task in 'run_d' failed with exit status 255:
        {task 809298864: strip run_d -> }
['/usr/vac/bin/xlc_r', '-X32', 'run_d']
ERROR: Failed compiling the bootloader. Please compile manually and rerun setup.py

```

**Comment by aixtools** [comment]
p.s. Would you be interested in looking at setting up my AIX server with CI testing? As something extra you could (optionally) do?

**Comment by bwoodsend** [comment]
Fixed by #4731 I believe? And are you offering some kind of CI testing platform we can use to ensure that PyInstaller works on AIX? I'm rather tempted to say yes...
